pkg/secrethub/internals/http/client.go

```diff
@@ -208,61 +207,106 @@ func (c *Client) GetUser(username string) (*api.User, error) {
 	out := &api.Repo{}
-	rawURL := fmt.Sprintf(pathRepos, c.base, namespace)
+	rawURL := fmt.Sprintf(pathRepos, c.base.String(), namespace)
 	err := c.post(rawURL, true, http.StatusCreated, in, out)
 	return out, errio.Error(err)
 }
 
 // GetRepoKeys retrieves the repo key of the user.
 func (c *Client) GetRepoKeys(namespace, repoName string) (*api.RepoKeys, error) {
 	out := &api.RepoKeys{}
-	rawURL := fmt.Sprintf(pathRepoKey, c.base, namespace, repoName)
+	rawURL := fmt.Sprintf(pathRepoKey, c.base.String(), namespace, repoName)
 	err := c.get(rawURL, true, out)
 	return out, errio.Error(err)
 }
 
 // DeleteRepo deletes a repo
 func (c *Client) DeleteRepo(namespace, repoName string) error {
-	rawURL := fmt.Sprintf(pathRepo, c.base, namespace, repoName)
+	rawURL := fmt.Sprintf(pathRepo, c.base.String(), namespace, repoName)
 	err := c.delete(rawURL, true, nil)
 	return errio.Error(err)
 }
 
 // AuditRepo gets the audit events for a given repo.
 func (c *Client) AuditRepo(namespace, repoName string, subjectTypes api.AuditSubjectTypeList) ([]*api.Audit, error) {
 	out := []*api.Audit{}
-	rawURL := fmt.Sprintf(pathRepoEvents, c.base, namespace, repoName)
-	query := make(url.Values)
+
+	requestURL := c.auditRepoURL(namespace, repoName)
+
 	if len(subjectTypes) > 0 {
-		query.Set("subject_types", subjectTypes.Join(","))
+		q := requestURL.Query()
+		q.Set("subject_types", subjectTypes.Join(","))
+		requestURL.RawQuery = q.Encode()
 	}
-	if len(query) > 0 {
-		rawURL += "?" + query.Encode()
+
+	err := c.get(requestURL.String(), true, &out)
+	return out, err
+}
+
+// AuditRepoPaginator returns a paginator to fetch the audit events for a given repo.
+func (c *Client) AuditRepoPaginator(namespace, repoName string) *AuditPaginator {
+	return newAuditPaginator(c.auditRepoURL(namespace, repoName), c)
+}
+
+func (c *Client) auditRepoURL(namespace, repoName string) url.URL {
+	return joinURL(c.base, fmt.Sprintf("/namespaces/%s/repos/%s/events", namespace, repoName))
+}
+
+func newAuditPaginator(requestURL url.URL, client *Client) *AuditPaginator {
+	return &AuditPaginator{
+		fetchPage: func(target *[]api.Audit, requestURL url.URL) error {
+			return client.get(requestURL.String(), true, &target)
+		},
+		requestURL: requestURL,
 	}
-	err := c.get(rawURL, true, &out)
-	return out, errio.Error(err)
+}
+
+type AuditPaginator struct {
+	fetchPage  func(target *[]api.Audit, requestURL url.URL) error
+	requestURL url.URL
+}
+
+func (pag *AuditPaginator) Next() ([]interface{}, error) {
+	events := make([]api.Audit, 50)
+
+	err := pag.fetchPage(&events, pag.requestURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(events) > 0 {
+		q := pag.requestURL.Query()
+		q.Set("starting_after", events[len(events)-1].EventID.String())
+		pag.requestURL.RawQuery = q.Encode()
+	}
+
+	res := make([]interface{}, len(events))
+	for i, event := range events {
+		res[i] = event
+	}
+	return res, nil
```

Comment:
I'm not a big fan of what happens here. For example, it does not look very extendable to other types. But I am okay with leaving it as is and refactoring it when we implement other paginators. Especially since it is internal code anyway.

Reply:
I believe my idea when implementing this was about the same; When we go for another paginator, I'd like to have the link header, so that all paginators determine the next page URL in the same way. Then, there's more that's the same for all paginators and we can extract that all together.